File: a105_cobra_driver/src/cmdvel_to_cobra.cpp
// src/cmdvel_simple_cobra.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <nav_msgs/msg/odometry.hpp>

#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

#include <string>
#include <thread>
#include <mutex>
#include <vector>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <chrono>
#include <cerrno>
#include <cstring>

class CmdVelToCobra : public rclcpp::Node {
public:
  CmdVelToCobra() : Node("cmdvel_to_cobra") {
    port_  = declare_parameter<std::string>("port", "/dev/ttyACM1");
    baud_  = declare_parameter<int>("baud", 115200);
    track_ = declare_parameter<double>("track", 0.228);
    tpm_   = declare_parameter<double>("tpm", 1000.0);
    rate_  = declare_parameter<double>("rate", 10.0);

    fd_ = open_serial(port_.c_str(), baud_);
    if (fd_ < 0) throw std::runtime_error("serial open failed");
    RCLCPP_INFO(get_logger(), "Opened %s @ %d", port_.c_str(), baud_);
    ::tcflush(fd_, TCIFLUSH);

    // Инициализация контроллера
    send_json("{\"T\":131,\"cmd\":1}");   // telemetry on
    send_json("{\"T\":142,\"cmd\":200}"); // 200 ms period
    send_json("{\"T\":143,\"cmd\":0}");   // echo off
    send_json("{\"T\":12}");              // enable motors
    send_json("{\"T\":3,\"lineNum\":0,\"Text\":\":-)\"}");
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // ROS I/O
    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("odom", 10);
    sub_ = create_subscription<geometry_msgs::msg::Twist>(
      "cmd_vel", 10, std::bind(&CmdVelToCobra::cmd_cb, this, std::placeholders::_1));

    timer_ = create_wall_timer(
      std::chrono::milliseconds(int(1000.0 / rate_)),
      std::bind(&CmdVelToCobra::send_cmd, this));

    rx_thread_ = std::thread([this]{ rx_loop(); });
  }

  ~CmdVelToCobra() override {
    run_rx_ = false;
    if (rx_thread_.joinable()) rx_thread_.join();
    if (fd_ >= 0) ::close(fd_);
  }

private:
  // --- Serial setup ---
  static speed_t baud_to_speed(int b) {
    switch(b){
      case 9600: return B9600;
      case 19200: return B19200;
      case 38400: return B38400;
      case 57600: return B57600;
      case 115200: return B115200;
      default: return B115200;
    }
  }

  int open_serial(const char* dev, int baud){
    int fd = ::open(dev, O_RDWR | O_NOCTTY | O_NONBLOCK);
    if (fd < 0) {
      RCLCPP_ERROR(get_logger(), "open(%s): %s", dev, std::strerror(errno));
      return -1;
    }
    fcntl(fd, F_SETFL, 0);
    termios tio{};
    if (tcgetattr(fd, &tio) != 0) { ::close(fd); return -1; }
    cfmakeraw(&tio);
    tio.c_cflag |= (CLOCAL | CREAD);
    cfsetispeed(&tio, baud_to_speed(baud));
    cfsetospeed(&tio, baud_to_speed(baud));
    tio.c_cc[VMIN] = 0; tio.c_cc[VTIME] = 1; // 100 ms
    if (tcsetattr(fd, TCSANOW, &tio) != 0) { ::close(fd); return -1; }
    return fd;
  }

  void send_json(const std::string& s){
    ::write(fd_, s.data(), s.size());
    ::write(fd_, "\r\n", 2); // контроллер принимает CRLF
    RCLCPP_INFO(get_logger(), "[TX] %s", s.c_str());
  }

  // --- Helpers: JSON ---
  static bool extract_int(const std::string& s, const char* key, int &out){
    std::string p = std::string("\"")+key+"\":";
    size_t i = s.find(p);
    if (i == std::string::npos) return false;
    i += p.size();
    out = std::atoi(s.c_str()+i);
    return true;
  }

  void drain_json_objects(std::string &buf, std::vector<std::string> &out){
    bool in_str = false, esc = false;
    int depth = 0;
    size_t start = std::string::npos;
    for (size_t i=0; i<buf.size(); ++i){
      char c = buf[i];
      if (in_str){
        if (esc) esc = false;
        else if (c == '\\') esc = true;
        else if (c == '"') in_str = false;
        continue;
      }
      if (c == '"'){ in_str = true; continue; }
      if (c == '{'){
        if (depth == 0) start = i;
        ++depth;
      } else if (c == '}'){
        --depth;
        if (depth == 0 && start != std::string::npos){
          out.emplace_back(buf.substr(start, i - start + 1));
          start = std::string::npos;
        }
      }
    }
    if (!out.empty()){
      const std::string &last = out.back();
      size_t endpos = buf.rfind(last);
      if (endpos != std::string::npos) buf.erase(0, endpos + last.size());
    } else if (buf.size() > 65536){
      buf.erase(0, buf.size() - 1024);
    }
  }

  rclcpp::Time now(){ return this->get_clock()->now(); }

  // --- ROS cmd_vel ---
  void cmd_cb(const geometry_msgs::msg::Twist::SharedPtr msg){
    std::lock_guard<std::mutex> lk(m_);
    last_v_ = msg->linear.x;
    last_w_ = msg->angular.z;
  }

  void send_cmd(){
    double v, w;
    { std::lock_guard<std::mutex> lk(m_); v = last_v_; w = last_w_; }
    char buf[96];
    std::snprintf(buf, sizeof(buf), "{\"T\":13,\"X\":%.3f,\"Z\":%.3f}", v, w);
    send_json(buf);
  }

  // --- RX ---
  void rx_loop(){
    std::string buf; buf.reserve(8192);
    auto last_t = now();
    while (run_rx_){
      char tmp[512];
      ssize_t n = ::read(fd_, tmp, sizeof(tmp));
      if (n > 0){
        buf.append(tmp, tmp + n);
        std::vector<std::string> objs;
        drain_json_objects(buf, objs);
        for (auto &js : objs) handle_json(js, last_t);
      } else {
        std::this_thread::sleep_for(std::chrono::milliseconds(5));
      }
    }
  }

  void handle_json(const std::string& js, rclcpp::Time &last_t){
    RCLCPP_INFO(get_logger(), "[RX] %s", js.c_str());

    int t=0;
    if (!extract_int(js, "T", t)) return;
    if (t != 1001) return;

    int odl=0, odr=0;
    if (!extract_int(js, "odl", odl)) return;
    if (!extract_int(js, "odr", odr)) return;

    auto tnow = now();
    if (!odom_ready_){
      last_left_ = odl; last_right_ = odr;
      odom_ready_ = true; last_t = tnow;
      return;
    }

    double dt = (tnow - last_t).seconds();
    if (dt <= 0) return;

    double Dl = (odl - last_left_) / tpm_;
    double Dr = (odr - last_right_) / tpm_;
    last_left_ = odl; last_right_ = odr;

    double dS  = 0.5*(Dl+Dr);
    double dTh = (Dr-Dl)/track_;
    double th_mid = yaw_ + 0.5*dTh;
    x_ += dS*std::cos(th_mid);
    y_ += dS*std::sin(th_mid);
    yaw_ += dTh;

    nav_msgs::msg::Odometry od;
    od.header.stamp = tnow;
    od.header.frame_id = "odom";
    od.child_frame_id  = "base_link";
    od.pose.pose.position.x = x_;
    od.pose.pose.position.y = y_;
    double half = 0.5*yaw_;
    od.pose.pose.orientation.z = std::sin(half);
    od.pose.pose.orientation.w = std::cos(half);
    od.twist.twist.linear.x  = dS/dt;
    od.twist.twist.angular.z = dTh/dt;
    odom_pub_->publish(od);

    last_t = tnow;
  }

  // --- members ---
  int fd_{-1};
  std::string port_;
  int baud_{115200};
  double track_{0.228}, tpm_{1000.0}, rate_{10.0};

  std::mutex m_;
  double last_v_{0.0}, last_w_{0.0};

  std::thread rx_thread_;
  bool run_rx_{true};

  bool odom_ready_{false};
  int last_left_{0}, last_right_{0};
  double x_{0}, y_{0}, yaw_{0};

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_;
  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv){
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CmdVelToCobra>());
  rclcpp::shutdown();
  return 0;
}

File: a105_cobra_py/a105_cobra_py/cmdvel_to_cobra.py
#!/usr/bin/env python3
import json, math, threading, time, serial
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry

class CmdVelToCobra(Node):
    def __init__(self):
        super().__init__('cmdvel_to_cobra_py')

        # params
        self.declare_parameter('port', '/dev/ttyACM1')
        self.declare_parameter('baud', 115200)
        self.declare_parameter('track', 0.228)   # m
        self.declare_parameter('tpm', 1000.0)    # ticks per meter
        self.declare_parameter('rate', 10.0)     # Hz, send T=13
        self.declare_parameter('poll_hz', 5.0)   # Hz, send T=130
        self.declare_parameter('odom_frame', 'odom')
        self.declare_parameter('base_frame', 'base_link')

        self.port = self.get_parameter('port').value
        self.baud = int(self.get_parameter('baud').value)
        self.track = float(self.get_parameter('track').value)
        self.tpm = float(self.get_parameter('tpm').value)
        self.rate = float(self.get_parameter('rate').value)
        self.poll_hz = float(self.get_parameter('poll_hz').value)
        self.odom_frame = self.get_parameter('odom_frame').value
        self.base_frame = self.get_parameter('base_frame').value

        # serial
        self.ser = serial.Serial(self.port, self.baud, timeout=0.05)
        self.get_logger().info(f"Serial open {self.port}@{self.baud}")

        # init controller (CRLF)
        self._jsend({"T":131, "cmd":1})   # telemetry on
        self._jsend({"T":142, "cmd":200}) # telemetry period ms
        self._jsend({"T":143, "cmd":0})   # echo off
        self._jsend({"T":12})             # motors enable
        self._jsend({"T":3, "lineNum":0, "Text":":-)"})
        time.sleep(0.1)

        # ROS IO
        self.sub = self.create_subscription(Twist, 'cmd_vel', self._cmd_cb, 10)
        self.pub = self.create_publisher(Odometry, 'odom', 10)

        self.timer_cmd  = self.create_timer(1.0/max(1e-6,self.rate), self._tick_cmd)
        self.timer_poll = self.create_timer(1.0/max(1e-6,self.poll_hz), self._tick_poll)

        # RX thread
        self._rx_run = True
        self._rx_th = threading.Thread(target=self._rx_loop, daemon=True)
        self._rx_th.start()

        # state
        self._v = 0.0
        self._w = 0.0
        self._last_left = 0
        self._last_right = 0
        self._odom_ready = False
        self._x = 0.0
        self._y = 0.0
        self._yaw = 0.0
        self._last_time = self.get_clock().now()

    # ---- helpers ----
    def _jsend(self, obj):
        s = json.dumps(obj)
        self.ser.write((s + "\r\n").encode('utf-8'))
        self.get_logger().info(f"[TX] {s}")

    def _cmd_cb(self, msg: Twist):
        self._v = msg.linear.x
        self._w = msg.angular.z

    def _tick_cmd(self):
        self._jsend({"T":13, "X":self._v, "Z":self._w})

    def _tick_poll(self):
        self._jsend({"T":130})

    # ---- RX ----
    def _rx_loop(self):
        buf = ""
        while self._rx_run:
            try:
                chunk = self.ser.read(512)
                if not chunk:
                    continue
                buf += chunk.decode('utf-8', errors='ignore')
                # простая сборка по балансировке скобок
                while True:
                    start = buf.find('{')
                    if start < 0:
                        buf = ""  # мусор до '{'
                        break
                    end = buf.find('}', start)
                    if end < 0:
                        # ждём продолжение
                        if start > 0:
                            buf = buf[start:]  # обрежем мусор слева
                        break
                    js = buf[start:end+1]
                    buf = buf[end+1:]
                    self._handle_json(js)
            except Exception as e:
                self.get_logger().error(f"RX error: {e}")
                time.sleep(0.1)

    def _handle_json(self, js: str):
        self.get_logger().info(f"[RX] {js}")
        try:
            data = json.loads(js)
        except Exception:
            return
        if data.get("T") != 1001:
            return

        # поддерживаем два формата: абсолюты (odl/odr) или дельты (dl/dr)
        tnow = self.get_clock().now()
        dt = (tnow - self._last_time).nanoseconds / 1e9
        if dt < 0:
            self._last_time = tnow
            return

        Dl = Dr = None
        if "odl" in data and "odr" in data:
            odl = int(data["odl"]); odr = int(data["odr"])
            if not self._odom_ready:
                self._last_left = odl; self._last_right = odr
                self._odom_ready = True
                self._last_time = tnow
                return
            Dl = (odl - self._last_left) / max(1e-6, self.tpm)
            Dr = (odr - self._last_right) / max(1e-6, self.tpm)
            self._last_left = odl; self._last_right = odr
        elif "dl" in data and "dr" in data:
            Dl = float(data["dl"]) / max(1e-6, self.tpm)
            Dr = float(data["dr"]) / max(1e-6, self.tpm)
            if not self._odom_ready:
                self._odom_ready = True

        if Dl is None or Dr is None:
            return
        if dt == 0.0:
            self._last_time = tnow
            return

        dS  = 0.5 * (Dl + Dr)
        dTh = (Dr - Dl) / max(1e-9, self.track)
        thm = self._yaw + 0.5*dTh
        self._x   += dS * math.cos(thm)
        self._y   += dS * math.sin(thm)
        self._yaw += dTh

        od = Odometry()
        od.header.stamp = tnow.to_msg()
        od.header.frame_id = self.odom_frame
        od.child_frame_id  = self.base_frame
        od.pose.pose.position.x = self._x
        od.pose.pose.position.y = self._y
        od.pose.pose.orientation.z = math.sin(self._yaw/2.0)
        od.pose.pose.orientation.w = math.cos(self._yaw/2.0)
        od.twist.twist.linear.x  = dS / dt
        od.twist.twist.angular.z = dTh / dt
        self.pub.publish(od)

        self.get_logger().info(f"[ODOM] dt={dt:.3f} Dl={Dl:.4f} Dr={Dr:.4f} x={self._x:.3f} y={self._y:.3f} yaw={self._yaw:.3f}")
        self._last_time = tnow


def main():
    rclpy.init()
    node = CmdVelToCobra()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node._rx_run = False
    time.sleep(0.1)
    node.destroy_node()
    rclpy.shutdown()
